src/Oauth2Client.php

```diff
@@ -0,0 +1,320 @@
+    curl_setopt($process, CURLOPT_RETURNTRANSFER, true);
+    $result = curl_exec($process);
+    curl_close($process);
+
+    return json_decode($result);
+
+  }
+
+
+  /**
+   * Grant Type 2: Authorization code grant
+   * Step 1 - generate a authorization code
+   *
+   *
+   * If you want to use this app for a different Shapeways User account.
+   *
+   *
+   * @link https://developers.shapeways.come/quick-start#authenticate
+   *
+   * redirect_uri - This determines where authentication requests will be sent and received by your app.
+   * Users will be redirected to this URL when they attempt to use your app.
+   */
+  public function generateAccessTokenAuthorizationGrant()
+  {
+    $clientId = $this->clientId;
+    $url = $this->baseApiUrl . "/oauth2/authorize?response_type=code&client_id=" . $clientId . "&redirect_uri=" . rawurlencode($this->redirectUrl);
+    header('Location: ' . $url);
+  }
+
+
+  /**
+   * Grant Type 2: Authorization code grant
+   * Step 2 - Generate access token
+   *
+   * @link https://developers.shapeways.com/quick-start#authenticate
+   *  function for handling call back of generateAccessTokenAuthorizationGrant() call above
+   *
+   * Use "access_token" from result for other API calls
+   *
+   * @return array - json decoded api response
+   */
+  public function handleAuthorizationGrantCallback()
+  {
+    $code = $_REQUEST['code'] ?? null;
+    $redirectURL = $this->redirectUrl;
+
+    if ($code === null) {
+      echo "Missing authorization code";
+      exit();
+    }
+    $url = $this->baseApiUrl . '/oauth2/token';
+
+    $data = array(
+      "grant_type" => "authorization_code",
+      "code" => $code,
+      "client_id" => $this->clientId,
+      "client_secret" => $this->clientSecret,
+      "redirect_uri" => $redirectURL
+    );
+
+    $ch = curl_init($url);
+    curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);
+    curl_setopt($ch, CURLOPT_POST, 1);
+    curl_setopt($ch, CURLOPT_POSTFIELDS, $data);
+
+    $result = curl_exec($ch);
+
+    curl_close($ch);
+    return json_decode($result);
+  }
+
+  /**
+   * Upload a model for the user
+   *
+   * @link https://developers.shapeways.com/api-reference#Models
+   *
+   * @param array $params the model data to set
+   * @return array the json response from the api call
+   * @throws ParameterValidationException
+   */
+  public function uploadModel($params)
+  {
+    $required = array('file', 'fileName', 'hasRightsToModel', 'acceptTermsAndConditions');
+    foreach ($required as $key) {
+      if (!array_key_exists($key, $params)) {
+        throw new ParameterValidationException('Shapeways\Oauth2CurlClient::addModel missing required key: ' . $key);
+      }
+    }
+
+    $params['file'] = rawurlencode(base64_encode($params['file']));
+
+    // json encode
+    $postData = json_encode($params);
+
+    $url = $this->baseApiUrl . '/models/v1';
+
+    $ch = curl_init($url);
+    curl_setopt($ch, CURLOPT_HTTPHEADER,
```

Comment:
It would be cool to use a PSR-7 compliant Request/Response library, such as Guzzle, in order to enable adoption in more mainstream PHP frameworks.

Reply:
^ this would be better than moving curl into its own function